### DISCUSSION TITLE
Alpha Diversity question

### DISCUSSION THREAD

**Discussion by RachBioHaz** [cafferychen777/MicrobiomeStat#19] [Q&A] [answered]
Hi, I was looking over the alpha diversity example and you state that if rarefying isn't used as a normalisation step, precalculated alpha diversity indices can be passed to the alpha.obj parameter in the alpha diversity analysis functions. What format does this need to be in and is there an example I can follow that does not use rarefying? Thanks

**Comment by cafferychen777** [comment, marked as answer]
Hi there,

Thank you for reaching out with your question about using precalculated alpha diversity indices with our R package's alpha diversity analysis functions.

If you choose not to rarefy your data as a normalization step, you can indeed pass precalculated alpha diversity indices to the `alpha.obj` parameter. 

Here is an example of how you can calculate the alpha diversity indices and then pass them to the `alpha.obj` parameter:

```R
# Calculate alpha diversity indices
alpha.obj <- mStat_calculate_alpha_diversity(
  x = data.obj$feature.tab,
  alpha.name = c("shannon", "simpson", "observed_species", "chao1", "ace", "pielou")
)

# Now, you can pass this `alpha.obj` directly to the alpha diversity analysis function
# For example, if you have a function `alpha_diversity_analysis`, you would do:
result <- mStat_alpha_xxxx(alpha.obj = alpha.obj, ...)

# Replace '...' with any other arguments your function requires.
```

Please make sure that the column names of `alpha.obj` match the expected diversity indices that the analysis function is looking for. If you need further assistance or have any more questions, feel free to ask.

Best regards,

Chen YANG

**Comment by adahal123** [comment]
Following up on this...
I created an alpha.obj with shannon and observed_species. It works great for **generate_alpha_trend_test_long**  and **generate_alpha_volatility_test_long**. 

However, when using **generate_alpha_test_long**, it still defaults to rarefy. There is no difference between using it and not using it. 

alpha_test <- generate_alpha_test_long(
  data.obj = ps.obj,
  alpha.obj = alpha.obj,
  alpha.name = c("shannon","observed_species"),
  time.var = "visit",
  t0.level = unique(ps.obj$meta.dat$visit)[1],
  ts.levels = unique(ps.obj$meta.dat$visit)[-1],
  group.var = "race",
  adj.vars = "sex"
)



...
Updated feature table to match the subsetted data.
Updated feature annotation to match the subsetted data.
Data subsetting complete. Returning updated data object.
Diversity analysis needs rarefaction! Call 'mStat_rarefy_data' to rarefy the data!
No depth specified, using the smallest column sum:  490408
Data has been subsetted based on the provided samIDs.

Warning messages:
1: In vegan::rrarefy(t(otu_tab), sample = depth) :
  function should be used for observed counts, but smallest count is 11
2: In vegan::rrarefy(t(otu_tab), sample = depth) :
  function should be used for observed counts, but smallest count is 9
3: In vegan::rrarefy(t(otu_tab), sample = depth) :
  function should be used for observed counts, but smallest count is 11


What am I doing wrong?

**Reply by cafferychen777** [reply]
Hi @adahal123,

Thank you for bringing this to my attention. I have noted the issue you've encountered with `generate_alpha_test_long` defaulting to rarefy even when precalculated alpha diversity indices are provided. I appreciate your detailed description and example, which will be helpful for diagnosing the problem.

I will investigate this issue and work on a fix as soon as possible. Once the fix is implemented, I will update the package and notify you of the new version.

In the meantime, if you have any further questions or encounter additional issues, please don't hesitate to reach out.

Best regards,
Caffery Yang

**Reply by cafferychen777** [reply]
Hi @adahal123,

I'm pleased to inform you that I have fixed the bug related to the `generate_alpha_test_long` function defaulting to rarefy even when precalculated alpha diversity indices are provided. The updated version of the MicrobiomeStat package has been uploaded.

Please give it a few hours for the changes to propagate, and then you can download the updated package. This should resolve the issue you encountered.

Thank you for your patience, and if you have any further questions or issues, feel free to reach out.

Best regards,
Caffery Yang

**Reply by adahal123** [reply]
Thank you for getting to it so quickly!!! 

This is slightly unrelated and might be an issue with my R studio specifically but when I try to use multiple covariates, I get the below warning message and the results seem to only take into account the first variable (see below). See warning message below. I can't figure out how to get around this issue. This happens for all functions that I use. 

alpha_trend <- generate_alpha_trend_test_long(
  data.obj = ps.obj,
  alpha.obj = alpha.obj,
  alpha.name = c("shannon")
  time.var = "visit",
  subject.var = "pid",
  group.var = "race",
  adj.vars = c("season", "sex", "severity") #this does not work...grrrrr... just does the first variable
)

...
The trend test in 'generate_alpha_trend_test_long' relies on a numeric time variable.
Please ensure that your time variable is coded as numeric.
If the time variable is not numeric, it may cause issues in computing the results of the trend test.
The time variable will be converted to numeric within the function if needed.
boundary (singular) fit: see help('isSingular')
boundary (singular) fit: see help('isSingular')
**Warning messages:
1: Using formula(x) is deprecated when x is a character vector of length > 1.
  Consider formula(paste(x, collapse = " ")) instead. 
2: Using formula(x) is deprecated when x is a character vector of length > 1.
  Consider formula(paste(x, collapse = " ")) instead.**

**Reply by cafferychen777** [reply]
Hi adahal123,

Thank you for your feedback and for using the MicrobiomeStat package. I appreciate your patience as we work through these issues.

Regarding the issue with the `generate_alpha_test_long` function, could you please provide a more detailed description of the problem you're encountering? It would be especially helpful if you could include a screenshot of the results after running the code. This will help me better understand the issue and provide a more accurate solution.

For the warning message related to multiple covariates in `generate_alpha_trend_test_long`, I'm currently investigating this and will update you as soon as I have a solution. In the meantime, if you have any additional information or examples that could help, please feel free to share.

Thank you for your cooperation and for helping improve the MicrobiomeStat package.

Best regards,
Caffery Yang

**Reply by adahal123** [reply]
This is the new error with the _generate_alpha_test_long_ it seems after you have updated it: 

Code:
alpha_test <- generate_alpha_test_long(
data.obj = ps.obj,
alpha.obj = alpha.obj,
alpha.name = c("shannon","observed_species"),
time.var = "visit",
t0.level = unique(ps.obj$meta.dat$visit)[1],
ts.levels = unique(ps.obj$meta.dat$visit)[-1],
group.var = "race",
adj.vars = "sex"
)


Result:
Updated feature table to match the subsetted data.
Updated feature annotation to match the subsetted data.
Data subsetting complete. Returning updated data object.
Data has been subsetted based on the provided condition.
Updated metadata to match the subsetted data.
The following samples were excluded: 
Updated feature table to match the subsetted data.
Updated feature annotation to match the subsetted data.
Data subsetting complete. Returning updated data object.
**Error in names(test.list) <- alpha.name : 
  'names' attribute [2] must be the same length as the vector [1]**


As I had mentioned, this alpha.obj has worked in other longitudinal alpha function tests.

**Reply by cafferychen777** [reply]
Hi @adahal123,

Thank you for reporting this issue with the generate_alpha_test_long function. I took a closer look and it appears the error may be occurring because the alpha diversity metrics specified in alpha.name when creating the alpha.obj object do not match the metrics specified in alpha.name when calling generate_alpha_test_long. 

The error message suggests the names attribute of the test.list object has length 1 while the alpha.name vector has length 2. This likely means the alpha.obj only contains a single diversity metric (either Shannon or observed species) while you are attempting to analyze both metrics by passing c("shannon","observed_species") to the alpha.name parameter in generate_alpha_test_long.

Can you please double check that the alpha metrics stored in alpha.obj match the ones you specify in the generate_alpha_test_long call? The alpha.name arguments need to be consistent between the two.

Let me know if you have any other questions! I appreciate you taking the time to test out the package and provide this helpful feedback.

Best regards,
Caffery

**Reply by cafferychen777** [reply]
Hi @adahal123,

I've fixed the issue in the `construct_formula` function where only the first variable in `adj.vars` was being included in the analysis when using multiple covariates. The updated package will be available for download in a few hours. Thank you for bringing this to my attention!

Best,
Chen YANG

**Reply by adahal123** [reply]
Thanks, multiple adj.vars works great now. 

As for the `generate_alpha_test_long`, the metrics "shannon" and "observed_species" are in the alpha.obj. The same alpha.obj and alpha.name = c("shannon", "observed_species") works for all other alpha diversity measure except for that one. For example:

**DOES NOT WORK**
```
alpha_test <- generate_alpha_test_long(
  data.obj = ps.obj,
  alpha.obj = alpha.obj,
  alpha.name = c("shannon", "observed_species"),
  time.var = "visit",
  t0.level = 1,
  ts.levels = c("2", "3"),
  group.var = "race",
  adj.vars ="sex"
)
```

Results:
Rule 1 passed: data.obj is a list.
Rule 2 passed: meta.dat is a data.frame.
Rule 3 passed: The row names of feature.tab match the row names of feature.ann.
Rule 4 passed: The column names of feature.tab match the row names of meta.dat.
Rule 5 passed: feature.tab is a matrix.
Rule 6 passed: feature.ann is a matrix.
Please note: The data components should follow base R data.frame and matrix structures, not phyloseq's formal class.
Validation passed.
Note: Passing validation does not guarantee the absence of all data issues. Further data exploration may be needed.
Data has been subsetted based on the provided condition.
Updated metadata to match the subsetted data.
The following samples were excluded: [SAMPLES from visit 2 and visit 3]
Updated feature table to match the subsetted data.
Updated feature annotation to match the subsetted data.
Data subsetting complete. Returning updated data object.
Data has been subsetted based on the provided condition.
Updated metadata to match the subsetted data.
The following samples were excluded: 
Updated feature table to match the subsetted data.
Updated feature annotation to match the subsetted data.
Data subsetting complete. Returning updated data object.
Error in names(test.list) <- alpha.name : 
  'names' attribute [2] must be the same length as the vector [1]

**WORKS**
```
alpha_test_change <- generate_alpha_change_test_long(
  data.obj = ps.obj,
  alpha.obj = alpha.obj,
  alpha.name = c("shannon", "observed_species"),
  time.var = "visit",
  t0.level = 1,
  ts.levels = c("2", "3"),
  subject.var = "pid",
  group.var = "race",
  adj.vars = "sex",
  alpha.change.func = "log fold change"
)
```
Results:
Data has been subsetted based on the provided samIDs.
Updated metadata to match the subsetted data.
The following samples were excluded: [SAMPLES from visit 3]
Updated feature table to match the subsetted data.
Updated feature annotation to match the subsetted data.
Data subsetting complete. Returning updated data object.
Data has been subsetted based on the provided samIDs.
Updated metadata to match the subsetted data.
The following samples were excluded: [SAMPLES from visit 2]
Updated feature table to match the subsetted data.
Updated feature annotation to match the subsetted data.
Data subsetting complete. Returning updated data object.

```
alpha_volatility <- generate_alpha_volatility_test_long(
  data.obj = ps.obj,
  alpha.obj = alpha.obj,
  alpha.name = c("shannon","observed_species"),
  time.var = "visit",
  subject.var = "pid",
  group.var = "race",
  adj.vars = "sex"
)
```

Result:
The volatility calculation in generate_alpha_volatility_test_long relies on a numeric time variable.
Please check that your time variable is coded as numeric.
If the time variable is not numeric, it may cause issues in computing the results of the volatility test.
You can ensure the time variable is numeric by mutating it in the metadata.
Alpha diversity has been adjusted for the following covariates: sex.
Alpha diversity has been adjusted for the following covariates: sex.

**Reply by cafferychen777** [reply]
I've resolved the issue with the `generate_alpha_test_long` function not working with the metrics "shannon" and "observed_species" in `alpha.obj`. The updated function should now work correctly. However, it's important to ensure that `t0.level` and `ts.levels` are consistent with the data type of the `time.var` in `data.obj$meta.dat`. They should be numeric if `time.var` is numeric, and character if `time.var` is character. Here's an example of how to use the updated function:

```r
alpha_test <- generate_alpha_test_long(
  data.obj = ps.obj,
  alpha.obj = alpha.obj,
  alpha.name = c("shannon", "observed_species"),
  time.var = "visit",
  t0.level = 1,  # Use numeric if "visit" is numeric in ps.obj$meta.dat
  ts.levels = c(2, 3),  # Use numeric if "visit" is numeric in ps.obj$meta.dat
  group.var = "race",
  adj.vars ="sex"
)
```

Please make sure to match the data type of `t0.level` and `ts.levels` with the data type of `time.var` in your dataset. If you encounter any further issues or have additional questions, feel free to reach out.
